File: Yoonseo Jeong/2023-05-06-binary-search.js
var search = function(nums, target) {
    if(!nums || target == undefined || nums.length < 1) return -1

    let startIndex = 0
    let lastIndex = nums.length-1

    while(startIndex <= lastIndex) {
        let midIndex = Math.floor((startIndex + lastIndex) / 2)
        if (target === nums[midIndex]) {
            return midIndex
        }
        else if (target < nums[midIndex]) {
            lastIndex = midIndex - 1
        }
        else if (target > nums[midIndex]) {
            startIndex = midIndex + 1
        }
    }

    return -1

    // complexity
    // time: O(log n)
    // space: O(1)
};

File: KwanyongJo/2023-05-06-Binary-Search.py
# Given an array of integers nums which is sorted in ascending order,
# and an integer target, write a function to search target in nums.
# If target exists, then return its index.
# Otherwise, return -1.
#
# You must write an algorithm with O(log n) runtime complexity.

class Solution:
    def search(self, nums: list[int], target: int) -> int:
        last_index = len(nums) -1
        start_index = 0

        while start_index <= last_index:
            mid_index = (last_index + start_index) //2

            if nums[mid_index] == target:
                return mid_index
            if nums[mid_index] < target:
                start_index = mid_index + 1
            else :
                last_index = mid_index -1


        return -1